### ISSUE TITLE
Spago appends packages.dhall to the Gitlab URL when using their API from packages.dhall

### ISSUE THREAD

**Issue by cscalfani** [purescript/spago#810]
I've set up an example of the problem at: https://gitlab.com/cscalfani/purescript-app.

REPRODUCE:

1. Clone the project.
2. Run `npm ci`.
3. Run `npx spago build`

The following will be displayed:

```
spago: 
Error: Remote file not found

HTTP status code: 404

URL: https://gitlab.com/api/v4/projects/28749897/repository/files/spago.dhall/packages.dhall?ref=main

Message:

1│ {"error":"404 Not Found"}
```

Notice that `packages.dhall` has been added to the end of the URL.  The original URL was:

```
https://gitlab.com/api/v4/projects/28749897/repository/files/spago.dhall/raw?ref=main
```

Gitlab requires the path suffix of `/raw` to get the file as text. Without it you get a JSON object with the contents in Base64 encoding.

Without the `/raw` and the suffix of the URL contains a filename and so `packages.dhall` will NOT be appended. But now, Gitlab will return JSON and not text data. And the base64 content key is unusable by Dhall.

The project that this "app" is referencing is a dummy library at URL:  https://gitlab.com/cscalfani/purescript-library if that helps.

**Comment by cscalfani** [comment]
I'm really confused how this is a Gitlab issue. Spago (or Dhall) should simply do an HTTP get with the supplied URL.

The fact that the URL is modified is the problem. This modification is outside of Gitlab's purview. Granted Gitlab's API sucks, but this appears to be a Dhall problem (or perhaps spago).

**Comment by toastal** [comment]
Can you not just reference it like `https://gitlab.com/cscalfani/purescript-app/-/raw/main/spago.dhall`? I thought most people just hard-copied in the dependencies anyhow. I've used GitLab before without much issue.

**Comment by f-f** [comment]
@cscalfani I'll try to expand on my answer above on why we can't really do much about this on Spago's side.

Everything that has to do with `.dhall` files in Spago is handled by the [Dhall compiler](https://dhall-lang.org/).
Dhall allows importing from URLs - that is, you can evaluate a file from the internet as part of evaluating a local expression.

Dhall implements certain parts of [RFC3986 section 5 resolution algorithm](https://tools.ietf.org/html/rfc3986#section-5) (see [Dhall's import system specification](https://github.com/dhall-lang/dhall-lang/blob/master/standard/imports.md) for more details): that is, it allows for relative references in files that are imported through a URL, and resolves the files according to the RFC above.

As an example, if we evaluate the following Dhall locally:
```dhall
https://example.com/some-file.dhall
```

...and that file has the following content:
```dhall
./some-other-file.dhall
```

...what we'll evaluate locally is really the content of `https://example.com/some-other-file.dhall`.

Now, the reasoning behind resolving imports in this way is the idea that if your relative imports work locally, and then you host your whole folder as is on a webserver, then your Dhall relative imports will still work when importing these files through a URL instead.

And this indeed works on GitHub repos: if you import a Dhall file, e.g. `https://raw.githubusercontent.com/purescript/spago/master/templates/spago.dhall`, that contains a relative reference (in this case `./packages.dhall`), that relative reference will still resolve correctly (in this case to `https://raw.githubusercontent.com/purescript/spago/master/templates/packages.dhall`).

GitLab does not follow this (really widespread and intuitive, at least to me) convention of having URLs reflect the local arrangement of files, but instead mangling the path when serving files in a different format.
[According to RFC7231](https://httpwg.org/specs/rfc7231.html#header.content-type) this should be achieved through the `Content-Type` header, which is why I'm claiming that this is a GitLab bug.

**Comment by toastal** [comment]
Non-ideal, but you _could_ use absolute URLs always. That's what I have done in the past -- mind you by accident because I hadn't bothered to think or try relative.

**Comment by cscalfani** [comment]
> Can you not just reference it like `https://gitlab.com/cscalfani/purescript-app/-/raw/main/spago.dhall`? I thought most people just hard-copied in the dependencies anyhow. I've used GitLab before without much issue.

Unfortunately, using this URL will redirect you to the logon for Gitlab.

**Comment by cscalfani** [comment]
f-f your explanation makes sense if I squint just right. Will have to ruminate on this a bit.

Thanks for explaining it to me.
